src/librustc_typeck/check/method/suggest.rs

```diff
@@ -61,12 +64,44 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
                 None);
 
             // If the item has the name of a field, give a help note
-            if let (&ty::TyStruct(did, _), Some(_)) = (&rcvr_ty.sty, rcvr_expr) {
+            if let (&ty::TyStruct(did, substs), Some(expr)) = (&rcvr_ty.sty, rcvr_expr) {
                 let fields = ty::lookup_struct_fields(cx, did);
-                if fields.iter().any(|f| f.name == item_name) {
-                    cx.sess.span_note(span,
-                        &format!("use `(s.{0})(...)` if you meant to call the \
-                                 function stored in the `{0}` field", item_ustring));
+
+                if let Some(field) = fields.iter().find(|f| f.name == item_name) {
+                    let expr_string = match cx.sess.codemap().span_to_snippet(expr.span) {
+                        Ok(expr_string) => expr_string,
+                        _ => "s".into() // Default to a generic placeholder for the
+                                        // expression when we can't generate a string
+                                        // snippet
+                    };
+
+                    // Determine if the field can be used as a function in some way
+                    let fn_once_trait_did = match cx.lang_items.require(FnOnceTraitLangItem) {
+                        Ok(trait_did) => trait_did,
+                        Err(err) => cx.sess.fatal(&err[..])
+                    };
+
+                    let field_ty = ty::lookup_field_type(cx, did, field.id, substs);
+                    let field_ty_substs = Substs::new_trait(vec![fcx.inh.infcx.next_ty_var()],
```

Comment:
Name could be misleading, the substitutions are for `FnOnce`, more specifically `vec![A]` in `FnOnce<A>`.